pageserver/src/layered_repository/inmemory_layer.rs

```diff
@@ -615,7 +619,13 @@ impl InMemoryLayer {
         // Figure out if we should create a delta layer, image layer, or both.
         let image_lsn: Option<Lsn>;
         let delta_end_lsn: Option<Lsn>;
-        if self.is_dropped() || !reconstruct_pages {
+        if self.is_dropped()
+            || !reconstruct_pages
+            || (self.seg.rel.is_blocky()
+                && self.get_seg_size(end_lsn_inclusive)? as u64 * BLCKSZ as u64
+                    > inner.page_versions.size() * 2
+                && n_delta_layers < MAX_DELTA_LAYERS)
```

Comment:
Ok, that makes sense, and I see how this PR is an improvement over the existing code. But..

`ImageLayer` stores a single LSN instead of an LSN for each page separately. So if the pageserver is doing garbage collection during this test, it will very likely need to read **from s3** to serve GetPage@LSN requests. Because the 11th random access to any segment will likely ask for a page older than the LSN that was garbage collected.

So I'd expect performance to be much worse than it is. Are we mocking s3 somehow in this test?

Reply:
Sorry, I do not completely understand you. GC in layered repo is done for layer, not WAL records or pages. This is one of the reasons why GC in layered repo is much faster than in buffered storage.

GC at pageserver is collecting its own data. It never needs to access S3 to perform GC or reconstruct page (get_page_at_lsn request).

Also  I do not understand why it should fail in this assertion.

Reply:
Ok I read some more code and I understand. I misunderstood how smgr works. I thought it asks for page at the LSN when **that** page was evicted, which would have been problematic. But smgr actually asks for page at the LSN when the **latest** page was evicted, so we're good :)

Reply:
Actually target LSN is max(last_lsn_written_by_smgr, last_lsn_received_by_pageserver).